### PR TITLE
Update psr-4-autoloader.md

### PR DESCRIPTION
**Pull request by AmyStephen**
Responses to your responses -- thanks Bernhard.

### CODE REVIEW

File: proposed/psr-4-autoloader/psr-4-autoloader.md [hunk 1]
```diff
@@ -68,22 +68,37 @@ for autoloader interoperability, by mapping namespaces to file system paths.
...
 
 4. Each namespace separator is converted to a `DIRECTORY_SEPARATOR` when
   loading from the file system.
 
   > Who does that? Must that happen or is it an option? What if I'm loading
   > from something else than the file system?
+  
+  >> Can you provide an example of this? IMO, this makes sense -- the interface
+  >> could certainly be a filesytem package [like](https://github.com/KnpLabs/Gaufrette)
+  >> or PHP filesystem commands.  An example of a limitation would help, though.
```
Comment (webmozart) [inline proposed/psr-4-autoloader/psr-4-autoloader.md:91]: Not sure what you mean by "example". We both know what the rule _is intended to say_, but it doesn't _actually say_ so. An improvement would be:

> The autoloader MUST convert each namespace separator to a `DIRECTORY_SEPARATOR` when loading from the file system.

But what if the autoloader caches the mapping? Then it will do the conversion in advance and check the mapping when loading. So we can further improve it:

> The autoloader MUST convert each namespace separator to a `DIRECTORY_SEPARATOR`.

But then the rule doesn't say what it applies to ("each namespace separator where?") and what the result is. So we can further refine it to:

> The autoloader MUST turn fully qualified class names into file system paths by replacing the namespace prefix with the corresponding base directory, replacing each namespace separator with a `DIRECTORY_SEPARATOR` and appending `.php` to the result.

Better. Now then the question is: How does the autoloader find the corresponding base directory? What if there are many?

This goes on.

To shorten my point: Many of the rules in this proposal have flaws like these. The inherent problem is that we are describing the _algorithm_ that should be implemented by an autoloader. What if an autoloader uses a different algorithm but produces the same results (file system paths) for the same inputs (FQCNs)? It won't be considered compliant.

That's why I turned the perspective of the proposal. I changed the rules to describe a _relationship_ ("correspondence") between namespaces and directories. How the autoloader establishes the relationship, whether it uses caching or any other fancy tricks, is secondary.

File: proposed/psr-4-autoloader/psr-4-autoloader.md [hunk 2]
```diff
@@ -68,22 +68,37 @@ for autoloader interoperability, by mapping namespaces to file system paths.
...
 
 5. The "fully qualified class name" MUST begin with a "namespace name", which
 MAY be followed by one or more additional namespace names, and MUST end in
 a class name.
 
   > Repetition of rule 1, which - as I have noted - is superfluous.
+  
+  >> Agree -- funny thing is -- you are driving to PSR-X -- Paul had the most
+  >> beautiful terse language but others added lots and lots of words. =)
+  >> This one _can_ go.
```
Comment (webmozart) [inline proposed/psr-4-autoloader/psr-4-autoloader.md:101]: I'm driving towards simplicity. Towards describing the contract that code directories ("libraries") must fulfill in order to be autoloadable. I'm convinced that nothing else is needed, but please prove me wrong.

File: proposed/psr-4-autoloader/psr-4-autoloader.md [hunk 3]
```diff
@@ -95,12 +110,22 @@ a class name.
   > * `Acme\Log\Formatter\`
   >
   > MUST correspond to one or more base directories? A pretty heavy restriction.
+  
+  >> No, but, LOL on that. Here's what I read that as -- any namespace prefix
+  >> identified above as those values mapped to base directories -- MUST
+  >> correspond to at least one base directory. It's an awkward wording, for sure.
+  >> So, what is it trying to say?  Basically, I think the intention is to say 
+  >> Namespace prefixes map to base directory(ies).
```
Comment (webmozart) [inline proposed/psr-4-autoloader/psr-4-autoloader.md:118]: You say "No", and we both know that this is not what the rule is intended to say. But it _says so_. Again we can improve the rule, but I repeat myself, because I already walked through this and the result can be seen in my proposal. :)

**Comment by webmozart** [comment]
I won't reply to each of your comments, because I think that I have made my point. The goal of my comments in the PSR was to show that the current proposal is flawed. You agreed in many cases.

What I did with the current proposal:
- check each rule for ambiguities
- remove them
- change them to not describe an algorithm ("transformation"), but a relationship

All of that in an iterative process, trying to keep the original _meaning_ of the PSR.

In other words, I did exactly the same thing that other people are doing right now, not "rewrite everything from scratch". [Compare the current proposal against the proposal under vote](https://github.com/php-fig/fig-standards/compare/250ea4b8bb780b48b1d0b58bddca63393f23c4e5...master#diff-4). It's just as "rewritten from scratch" (or not, depending on your POV) as the improvement I presented is.

File: proposed/psr-4-autoloader/psr-4-autoloader.md (hunk 1 above)
Comment (AmyStephen) [inline proposed/psr-4-autoloader/psr-4-autoloader.md:91]: Actually, I thought your point was the file was coming from something other than the file system. Totally did not understand what you meant, which is why I asked for an example.

I had made a similar point when discussing the transformation rules. It would be better to define what input is required to define a pair (i.e., a namespace prefix and a filesystem path) and the result of transformation (the prefix replaces the filesystem path; each subfolder is separated by a slash and the .php removed from the class file) -- but not delve into the transformation, itself.

Much of that wording has been cleaned up. Might be good to start with a new draft.

File: proposed/psr-4-autoloader/psr-4-autoloader.md (hunk 2 above)
Comment (AmyStephen) [inline proposed/psr-4-autoloader/psr-4-autoloader.md:101]: The mere fact that the new proposal limits itself to libraries is part of the problem, it's limiting the use already that PSR-4 enables. I've given many examples, Bernhard, of how your proposal imposes limits PSR-4 enables. Many examples.

**Comment by AmyStephen** [comment]
To be clear, the new approach imposes limits that PSR-4 does not have. I've identified those limits in the various _walls of text_ I shared. 

In sum: 
1) the scope of the new proposal limits the usefulness of the PSR to packages, only. Said differently, the ability to apply these rules (and use a compliant autoloader) for a CMS or distribution is gone.
2) overrides are not possible -- having multiple FQCN for test class files is not possible, nor is distributing multiple classfiles each with the same FQCN (this impacts unit testing, developers overriding files with private methods, and so on); 
3) the new proposal will create the same four empty folder structure that is inherent in PSR-0 for Composer users. 

The benefits for these limitations are not clear to me.

PSR-4 provides for all of those flexibilities and does so in an air-tight fashion. The only ambiguity that I see relates to the selection process for multiple class files with the same FQCN. That issue is under discussion.

It would not be possible for me to use your proposal with [Molajo](https://github.com/Molajo). I would have to opt for non-compliance due to the limitations imposed. I've worked with PSR-4 for months now with code. It's an excellent standard. I'm fine with changing the language in favor of clarity. I object to reducing any flexibility absent a clear reason and example for requiring such be imposed. I've not seen those reasons in these changes.

Purpose matters more than form, in the end.

**Comment by webmozart** [comment]
To answer your points:

1) I disagree. "Library" - I have explicitly defined this as a generic concept and alias for similar concepts, like "applications", "modules" etc. - applies to CMS's and distributions just as well.
2) I changed the wording. Overrides are possible now, even within the same "library".
3) This is not true. A namespace of any length (e.g. `\Foo\Bar\Baz`) may be mapped to any directory, which was not possible in PSR-0.

Please reread the current version of the proposal. I have integrated feedback from many people and really think that your declining stance is unjustified. :)

**Comment by AmyStephen** [comment]
@bschussek I don't mean to be unfair -- one question, then I just don't have time to review again until tomorrow. Did you remove the requirement to the src and test folder structure? If so, then I will need to look more closely.

**Comment by webmozart** [comment]
I think you have mistaken an example for a requirement there :) So for the sake of clarity: Everything written in block quotes is a _comment_ which won't be present in the final PSR.
